File: jmh-benchmarks/src/jmh/java/net/minestom/server/tag/TagWriteBenchmark.java
package net.minestom.server.tag;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Measurement;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.Warmup;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;

@Warmup(iterations = 5, time = 1000, timeUnit = TimeUnit.MILLISECONDS)
@Measurement(iterations = 10, time = 1000, timeUnit = TimeUnit.MILLISECONDS)
@Fork(3)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Benchmark)
public class TagWriteBenchmark {
    static final Tag<String> TAG = Tag.String("key");

    TagHandler tagHandler;
    Tag<String> secondTag;

    Map<String, String> map;
    Map<String, String> concurrentMap;

    @Setup
    public void setup() {
        // Tag benchmark
        this.tagHandler = TagHandler.newHandler();
        tagHandler.setTag(TAG, "value");
        secondTag = Tag.String("key");
        // Concurrent map benchmark
        map = new HashMap<>();
        map.put("key", "value");
        // Hash map benchmark
        concurrentMap = new ConcurrentHashMap<>();
        concurrentMap.put("key", "value");
    }

    @Benchmark
    public void writeConstantTag() {
        tagHandler.setTag(TAG, "value");
    }

    @Benchmark
    public void writeDifferentTag() {
        tagHandler.setTag(secondTag, "value");
    }

    @Benchmark
    public void writeNewTag() {
        tagHandler.setTag(Tag.String("key"), "value");
    }

    @Benchmark
    public void writeConcurrentMap() {
        concurrentMap.put("key", "value");
    }

    @Benchmark
    public void writeMap() {
        map.put("key", "value");
    }
}

File: jmh-benchmarks/src/jmh/java/net/minestom/server/tag/TagWriteBenchmark.kt
package net.minestom.server.tag

import org.jglrxavpok.hephaistos.nbt.mutable.MutableNBTCompound.set
import org.jglrxavpok.hephaistos.nbt.mutable.MutableNBTCompound.setString
import net.minestom.server.tag.TagHandler
import org.jglrxavpok.hephaistos.nbt.mutable.MutableNBTCompound
import net.minestom.server.tag.TagReadBenchmark
import java.util.concurrent.ConcurrentHashMap
import org.jglrxavpok.hephaistos.nbt.NBT
import org.openjdk.jmh.infra.Blackhole
import net.minestom.server.tag.TagWriteBenchmark
import java.util.function.IntFunction
import net.minestom.server.event.EventNode
import java.lang.Runnable
import net.minestom.server.timer.TaskSchedule
import java.util.concurrent.atomic.AtomicInteger
import net.minestom.server.instance.palette.Palette.EntrySupplier
import net.minestom.server.instance.palette.Palette.EntryConsumer
import org.openjdk.jmh.annotations.*
import java.util.concurrent.TimeUnit

@Warmup(iterations = 5, time = 1000, timeUnit = TimeUnit.MILLISECONDS)
@Measurement(iterations = 10, time = 1000, timeUnit = TimeUnit.MILLISECONDS)
@Fork(3)
@BenchmarkMode(
    Mode.AverageTime
)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Benchmark)
class TagWriteBenchmark {
    var tagHandler: TagHandler? = null
    var secondTag: Tag<String>? = null
    var compound: MutableNBTCompound? = null
    @Setup
    fun setup() {
        // Tag benchmark
        tagHandler = TagHandler.newHandler()
        tagHandler!!.setTag(TAG, "value")
        secondTag = Tag.String("key")
        // NBT benchmark
        compound = MutableNBTCompound(ConcurrentHashMap())
        compound!!["key"] = NBT.String("value")
    }

    @Benchmark
    fun writeConstantTag() {
        tagHandler!!.setTag(TAG, "value")
    }

    @Benchmark
    fun writeDifferentTag() {
        tagHandler!!.setTag(secondTag, "value")
    }

    @Benchmark
    fun writeNewTag() {
        tagHandler!!.setTag(Tag.String("key"), "value")
    }

    @Benchmark
    fun writeConstantTagFromCompound() {
        compound!!.setString("key", "value")
    }

    companion object {
        val TAG = Tag.String("key")
    }
}